### DISCUSSION TITLE
RSocket python over Websocket using aiohttp

### DISCUSSION THREAD

**Discussion by jell-o-fishi** [aio-libs/aiohttp#7110] [Show and tell]
[RSocket](https://rsocket.io/) - Application protocol providing Reactive Streams semantics.

I am the maintainer of [rsocket-py](https://github.com/rsocket/rsocket-py), a python implementation of the protocol.
There is integration with aiohttp using RSocket over Websocket, both server and client.

Would appreciate feedback.

Available on [pypi](https://pypi.org/project/rsocket/).

**Comment by Dreamsorcerer** [comment]
Feedback on implementation?

**Reply by jell-o-fishi** [reply]
Yes. Either usage or implementation. Anything really. There are example usages with aiohttp under the /examples folder of the main repository. There is a (preview of a) tutorial at https://rad-fox-9dcabd.netlify.app/guides/rsocket-py/tutorial/ for RSocket itself, the final step of which is using aiohttp instead of simple tcp transport.

**Reply by jell-o-fishi** [reply]
Specifically these:
examples:
https://github.com/rsocket/rsocket-py/blob/master/examples/server_aiohttp_websocket.py
https://github.com/rsocket/rsocket-py/blob/master/examples/client_websocket.py
RSocket transport adapter:
https://github.com/rsocket/rsocket-py/blob/master/rsocket/transports/aiohttp_websocket.py

**Reply by Dreamsorcerer** [reply]
OK, `await web._run_app(app, port=6565)`. This is clearly a private function and should not be used. I think for your example, you should just use the standard `web.run_app()` function (with no call to `asyncio.run()`). (https://docs.aiohttp.org/en/stable/web_quickstart.html#run-a-simple-web-server)

As you only add 1 handler, I'd also consider using `app.router.add_get()` as a slightly more concise method.

Applies to both:
https://github.com/rsocket/rsocket-py/blob/master/examples/tutorial/step8/chat_server.py
https://github.com/rsocket/rsocket-py/blob/master/examples/server_aiohttp_websocket.py

I also wonder if the Transport class could have `__aenter__()` and `__aexit__()` to automatically call `.connect()`/`.close()`.

Otherwise, looks like it should be good.

**Reply by jell-o-fishi** [reply]
Wouldn't it be reasonable to have a formal way to run the web server without it being in control of the event loop ? As part of another application ? For example in the tests, I had to use the private function to run it, because the test framework was already in charge of the event loop. And in the server_aiohttp_websocket.py I placed it inside asyncclick.  Am i missing something ?

**Reply by jell-o-fishi** [reply]
Regarding the asynccontextmanager I thought about that, but requires some major refactoring of the package. Maybe in the future.

**Reply by Dreamsorcerer** [reply]
> Wouldn't it be reasonable to have a formal way to run the web server without it being in control of the event loop?

https://docs.aiohttp.org/en/stable/web_advanced.html#application-runners

For most applications though, it's usually easier and more reliable to let aiohttp run the program, and hook other parts of your program into that (using on_startup/cleanup_ctx). This allows aiohttp to ensure those other parts are started/stopped alongside the web app.
See: https://docs.aiohttp.org/en/stable/web_advanced.html#complex-applications

> I had to use the private function to run it, because the test framework was already in charge of the event loop.

I've not looked at the tests, but generally you want to use the testing utils we provide, which should run the application for you:
https://docs.aiohttp.org/en/stable/testing.html

> And in the server_aiohttp_websocket.py I placed it inside asyncclick.

I'd suggest not using asynclick for that. Just use a regular function which calls `run_app()`. I'm not sure I'd want asyncclick to be controlling the event loop, I only see click as an entrypoint to start running the application (but, maybe you have something different in mind, I'm not familiar with the full capabilities of it).
